unit-tests/src/test/scala/java/nio/file/FilesSuite.scala

```diff
@@ -891,6 +894,64 @@ object FilesSuite extends tests.Suite {
     }
   }
 
+  // This test was inspired by Issue #1354
+  // "NIO File Walker fails on broken links."
+  test("Files.walkFileTree respects FileVisitOption.FOLLOW_LINKS") {
+    withTemporaryDirectory { dirFile =>
+      val dir = dirFile.toPath()
+      val f0  = dir.resolve("f0")
+      val f1  = dir.resolve("f1")
+
+      val brokenLink    = dir.resolve("brokenlink")
+      val missingTarget = dir.resolve("missingtarget")
+
+      Files.createFile(f0)
+      Files.createFile(f1)
+
+      assert(Files.exists(f0) && Files.isRegularFile(f0))
+      assert(Files.exists(f1) && Files.isRegularFile(f1))
+
+      Files.createFile(missingTarget)
+
+      assert(
+        Files.exists(missingTarget) &&
+          Files.isRegularFile(missingTarget))
+
+      // Create valid symbolic link from brokenLink to missingTarget,
+      // then remove missingTarget to break link.
+
+      Files.createSymbolicLink(brokenLink, missingTarget)
+
+      assert(Files.exists(brokenLink) && Files.isSymbolicLink(brokenLink),
+             "File brokenLink does not exist or is not a symbolic link.")
+
+      Files.delete(missingTarget)
+
+      assert(!Files.exists(missingTarget),
+             "File missingTarget should not exist.")
+
+      val visitor = new QueueingVisitor()
+
+      // will not follow links, so broken link does not throw exception.
+      Files.walkFileTree(dir, visitor)
+
+      val result   = visitor.length
+      val expected = 4
+
+      assert(result == expected, s"result: $result != expected: $expected")
```

Comment:
It would be better if you actually tested that the expected files were returned rather than just the length.

Reply:
Agreed. Good catch.  I must have written that at the end of a sprint...

Reply:
I changed the code as you suggested.  I now check that the set of files actually visited
matches the set of files expected.  I skip the pre and post-visit directory as that is not
useful.

From what I could find out, the order of visiting sub-directories is unspecified.  I could
find nothing, in limited time, that specified the order of files visited. I take it to be unspecified.
In practice, the order on SN appears to be reverse alphabetical.  I  created and compared
Sets to avoid relying upon any particular order of visitation.